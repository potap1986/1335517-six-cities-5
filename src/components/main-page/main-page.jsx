import React from 'react';
import PropTypes from 'prop-types';
import Map from '../map/map';
import Sort from '../sort/sort';
import OfferList from '../offer-list/offer-list';
import CityChange from '../city-change/city-change';
import EmptyMainPage from '../empty-main-page/empty-main-page';
import {connect} from 'react-redux';
import {ActionCreator} from '../../store/action';

const {changeCity, resetSortType, resetHoveredOffer, setCityOffers, /* resetCityOffers, loadOffers, */setHoveredOffer, setSortType} = ActionCreator;

const MainPage = (props) => {
  const {offers, hoveredOffer, city, onOfferClick, onCityChange, onSortTypeClick, onOfferHover} = props;

  return (
    <div className="page page--gray page--main">
      <header className="header">
        <div className="container">
          <div className="header__wrapper">
            <div className="header__left">
              <a className="header__logo-link header__logo-link--active">
                <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41"/>
              </a>
            </div>
            <nav className="header__nav">
              <ul className="header__nav-list">
                <li className="header__nav-item user">
                  <a className="header__nav-link header__nav-link--profile" href="#">
                    <div className="header__avatar-wrapper user__avatar-wrapper">
                    </div>
                    <span className="header__user-name user__name">Oliver.conner@gmail.com</span>
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </header>

      <main className="page__main page__main--index">
        <h1 className="visually-hidden">Cities</h1>
        <div className="tabs">

          <CityChange onCityChange={onCityChange} />

        </div>
        <div className="cities">
          {offers.length === 0
            ?
            <EmptyMainPage city={city}/>
            :
            <div className="cities__places-container container">
              <section className="cities__places places">
                <h2 className="visually-hidden">Places</h2>
                <b className="places__found">{`${offers.length} ${offers.length > 1 ? `places` : `place`}`} to stay in {city}</b>
                <Sort onSortTypeClick={onSortTypeClick}/>

                <OfferList onOfferClick={onOfferClick} onOfferHover={onOfferHover}/>

              </section>
              <div className="cities__right-section">
                <section className="cities__map map">
                  <Map offers={offers} hoveredOffer={hoveredOffer} />
                </section>
              </div>
            </div>
          }
        </div>
      </main>
    </div>
  );
};

MainPage.propTypes = {
  offers: PropTypes.array.isRequired,
  hoveredOffer: PropTypes.number.isRequired,
  city: PropTypes.string.isRequired,
  onOfferClick: PropTypes.func.isRequired,
  onCityChange: PropTypes.func.isRequired,
  onSortTypeClick: PropTypes.func.isRequired,
  onOfferHover: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  city: state.APPLICATION.currentCity,
  offers: state.APPLICATION.offersForCity,
  hoveredOffer: state.APPLICATION.hoveredOffer,
});

const mapDispatchToProps = (dispatch) => ({
  onCityChange: (city, allOffers) => {
    dispatch(resetSortType());
    dispatch(resetHoveredOffer());
    dispatch(changeCity(city));
    dispatch(setCityOffers(allOffers.filter((offer) => offer.hotelCity.name === city)));
  },
  onOfferHover(offer) {
    dispatch(setHoveredOffer(offer));
  },
  onSortTypeClick(sort) {
    dispatch(resetHoveredOffer());
    dispatch(setSortType(sort));
  },
});

export {MainPage};
export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
