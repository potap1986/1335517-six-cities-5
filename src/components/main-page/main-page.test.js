import renderer from 'react-test-renderer';
import {MainPage} from './main-page';

jest.mock(`../offer-list/offer-list`, () => `OfferList`);
jest.mock(`../map/map`, () => `Map`);
jest.mock(`../header/header`, () => `Header`);
jest.mock(`../city-change/city-change`, () => `CityChange`);
jest.mock(`../empty-main-page/empty-main-page`, () => `EmptyMainPage`);
jest.mock(`../sort/sort`, () => `Sort`);

jest.mock(`react-router-dom`, () => ({
  Link: () => null
}));

const offers = [{
  hotelCity: {
    name: `Lafayette`,
    location: {
      lat: 30,
      lng: 19,
      zoom: 16,
    },
  },
  previewImage: `mock`,
  images: [`mock`, `mock`],
  title: `mock`,
  isFavorite: true,
  isPremium: true,
  rating: 5,
  type: `hotel`,
  bedrooms: 2,
  adults: 2,
  price: 22,
  goods: [`mock`, `mock`],
  hotelHost: {
    id: 2,
    name: `John Doe`,
    isProUser: true,
    avatar: `mock`,
  },
  description: `Mock. Mock.`,
  coordinates: {
    lat: 50,
    lng: 50,
    zoom: 16,
  },
  id: 1,
},
{
  hotelCity: {
    name: `Lafayette`,
    location: {
      lat: 30,
      lng: 19,
      zoom: 16,
    },
  },
  previewImage: `mock`,
  images: [`mock`, `mock`],
  title: `mock`,
  isFavorite: true,
  isPremium: true,
  rating: 5,
  type: `hotel`,
  bedrooms: 2,
  adults: 2,
  price: 22,
  goods: [`mock`, `mock`],
  hotelHost: {
    id: 2,
    name: `John Doe`,
    isProUser: true,
    avatar: `mock`,
  },
  description: `Mock. Mock.`,
  coordinates: {
    lat: 50,
    lng: 50,
    zoom: 16,
  },
  id: 2,
}];

it(`Main screen correctly renders after relaunch`, () => {
  const mainScreen = renderer
    .create(<MainPage
      offers={offers}
      hoveredOffer={0}
      city={`Berlin`}
      onOfferClick={jest.fn()}
      onCityChange={jest.fn()}
      onSortTypeClick={jest.fn()}
      onOfferHover={jest.fn()}
    />)
  .toJSON();

  expect(mainScreen).toMatchSnapshot();
});
