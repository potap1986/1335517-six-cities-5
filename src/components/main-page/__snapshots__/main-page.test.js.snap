// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main screen correctly renders after relaunch 1`] = `
<div
  className="page page--gray page--main"
>
  <Header />
  <main
    className="page__main page__main--index"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="tabs"
    >
      <CityChange
        onCityChange={[MockFunction]}
      />
    </div>
    <div
      className="cities"
    >
      <div
        className="cities__places-container container"
      >
        <section
          className="cities__places places"
        >
          <h2
            className="visually-hidden"
          >
            Places
          </h2>
          <b
            className="places__found"
          >
            2 places
             to stay in 
            Berlin
          </b>
          <Sort
            onSortTypeClick={[MockFunction]}
          />
          <OfferList
            onOfferClick={[MockFunction]}
            onOfferHover={[MockFunction]}
          />
        </section>
        <div
          className="cities__right-section"
        >
          <section
            className="cities__map map"
          >
            <Map
              hoveredOffer={0}
              offers={
                Array [
                  Object {
                    "adults": 2,
                    "bedrooms": 2,
                    "coordinates": Object {
                      "lat": 50,
                      "lng": 50,
                      "zoom": 16,
                    },
                    "description": "Mock. Mock.",
                    "goods": Array [
                      "mock",
                      "mock",
                    ],
                    "hotelCity": Object {
                      "location": Object {
                        "lat": 30,
                        "lng": 19,
                        "zoom": 16,
                      },
                      "name": "Lafayette",
                    },
                    "hotelHost": Object {
                      "avatar": "mock",
                      "id": 2,
                      "isProUser": true,
                      "name": "John Doe",
                    },
                    "id": 1,
                    "images": Array [
                      "mock",
                      "mock",
                    ],
                    "isFavorite": true,
                    "isPremium": true,
                    "previewImage": "mock",
                    "price": 22,
                    "rating": 5,
                    "title": "mock",
                    "type": "hotel",
                  },
                  Object {
                    "adults": 2,
                    "bedrooms": 2,
                    "coordinates": Object {
                      "lat": 50,
                      "lng": 50,
                      "zoom": 16,
                    },
                    "description": "Mock. Mock.",
                    "goods": Array [
                      "mock",
                      "mock",
                    ],
                    "hotelCity": Object {
                      "location": Object {
                        "lat": 30,
                        "lng": 19,
                        "zoom": 16,
                      },
                      "name": "Lafayette",
                    },
                    "hotelHost": Object {
                      "avatar": "mock",
                      "id": 2,
                      "isProUser": true,
                      "name": "John Doe",
                    },
                    "id": 2,
                    "images": Array [
                      "mock",
                      "mock",
                    ],
                    "isFavorite": true,
                    "isPremium": true,
                    "previewImage": "mock",
                    "price": 22,
                    "rating": 5,
                    "title": "mock",
                    "type": "hotel",
                  },
                ]
              }
            />
          </section>
        </div>
      </div>
    </div>
  </main>
</div>
`;
